test_rmw_implementation/test/test_client.cpp

```diff
@@ -385,26 +385,23 @@ TEST_F(CLASSNAME(TestClientUse, RMW_IMPLEMENTATION), service_server_is_available
 TEST_F(CLASSNAME(TestClientUse, RMW_IMPLEMENTATION), service_server_is_available_good_args)
 {
   bool is_available = false;
-  rmw_ret_t ret = RMW_RET_ERROR;
-  SLEEP_AND_RETRY_UNTIL(rmw_intraprocess_discovery_delay, rmw_intraprocess_discovery_delay * 10) {
-    ret = rmw_service_server_is_available(node, client, &is_available);
-    if (RMW_RET_OK == ret && is_available) {
-      break;
-    }
-  }
-  EXPECT_EQ(ret, RMW_RET_OK) << rmw_get_error_string().str;
-  EXPECT_FALSE(is_available) << rmw_get_error_string().str;
+  rmw_ret_t ret = rmw_service_server_is_available(node, client, &is_available);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+  ASSERT_FALSE(is_available);
   rmw_reset_error();
 
   rmw_service_t * service = rmw_create_service(node, ts, service_name, &qos_profile);
   ASSERT_NE(nullptr, client) << rcutils_get_error_string().str;
+
+  is_available = false;
   SLEEP_AND_RETRY_UNTIL(rmw_intraprocess_discovery_delay, rmw_intraprocess_discovery_delay * 10) {
     ret = rmw_service_server_is_available(node, client, &is_available);
-    if (RMW_RET_OK == ret && is_available) {
+    EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+    rmw_reset_error();
+    if (is_available) {
```

Comment:
I'm not sure that this is the right logic here (I also don't think the previous logic is correct either).

What we are doing here is looping for a specified amount of time, waiting to see if the service server has become available.  From what I understand from the documentation, `rmw_service_server_is_available` should *always* return `RMW_RET_OK`, unless the typesupport didn't match or an unexpected error occurred.  But in those latter 2 cases, that is clearly a failure of this test, and so further waiting doesn't seem warranted.  Thus, I would think this logic would be:

```
if (ret != RMW_RET_OK) {
    break;
}
if (is_available) {
    break;
}
```

And then we keep both expectations below.  What do you think?

(this same comment applies to both loops below)

Reply:
I'd be fine with it. We could also just `ASSERT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;` right after calling `rmw_service_is_available`. Which one would you prefer?

Reply:
> I'd be fine with it. We could also just `ASSERT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;` right after calling `rmw_service_is_available`. Which one would you prefer?

The downside to doing an `ASSERT` is that we'd never hit the cleanup.  While that doesn't matter *too* much in a test, it has the side-effect of making clang static analysis complain, which I'd rather avoid.

Reply:
Oh, but that's because the test is not doing clean-up properly ! See 2da26ad311836667e68f98f4ab5c32d42876882a.